src/os/layer/preset/layerpresetmanager.js

```diff
@@ -1,223 +1,240 @@
- * @param {!os.events.LayerEvent} event The layer event.
- * @protected
- */
-os.layer.preset.LayerPresetManager.prototype.onLayerAdded = function(event) {
-  const layer = os.implements(event.layer, os.layer.ILayer.ID) ?
-      /** @type {os.layer.ILayer} */ (event.layer) : undefined;
-
-  if (layer) {
-    const layerId = layer.getId();
-    const layerOptions = layer.getLayerOptions();
-    if (layerId && layerOptions && layerOptions['applyDefaultPresets']) {
-      // Apply default presets to the layer when configured.
-      this.getPresets(layerId, true);
+  /**
+   * Handle layer added to the map.
+   * @param {!os.events.LayerEvent} event The layer event.
+   * @protected
+   */
+  onLayerAdded(event) {
+    const layer = osImplements(event.layer, ILayer.ID) ?
+        /** @type {ILayer} */ (event.layer) : undefined;
+
+    if (layer) {
+      const layerId = layer.getId();
+      const layerOptions = layer.getLayerOptions();
+      if (layerId && layerOptions && layerOptions['applyDefaultPresets']) {
+        // Apply default presets to the layer when configured.
+        this.getPresets(layerId, true);
+      }
     }
   }
-};
-
-
-/**
- * Gets a promise that resolves to the presets for a given layer type.
- *
- * @param {string} type
- * @param {string} url
- */
-os.layer.preset.LayerPresetManager.prototype.registerPreset = function(type, url) {
-  var promise = this.presets_[type];
-
-  if (!this.requested_[url]) {
-    this.requested_[url] = true;
-    var request = new os.net.Request(url);
-
-    if (promise) {
-      // presets are already loaded for that type, so attempt to load the new ones and replace the promise
-      promise.then(function(originalPresets) {
-        var newPromise = request.getPromise().then(function(result) {
-          var presets = /** @type {Array<osx.layer.Preset>} */ (JSON.parse(result));
-
-          if (presets && originalPresets) {
-            presets = presets.concat(originalPresets);
-          }
 
-          this.presets_[type] = newPromise;
-
-          return presets;
-        }, this.handleLoadError, this);
-      }, this.handleLoadError, this);
-    } else {
-      // no presets yet, so load them
-      this.presets_[type] = request.getPromise().then(function(result) {
-        var presets = /** @type {Array<osx.layer.Preset>} */ (JSON.parse(result));
-        return presets;
-      }, this.handleLoadError, this);
+  /**
+   * @param {boolean|undefined} b
+   * @return {!boolean}
+   */
+  isAdmin(b) {
+    if (b === true || b === false) {
+      this.admin_ = (b === true);
     }
+    return this.admin_;
   }
-};
 
+  /**
+   * @param {string} key
+   * @param {IPresetService} service
+   * @param {...} opt
+   * @return {boolean} true if prior value overwritten
+   */
+  registerService(key, service, ...opt) {
+    this.presets_ = {}; // set presets to reload every layer
+    return this.services_.register.apply(this.services_, [key, service].concat(opt));
```

Comment:
Does this need to be called with an `apply`?

Reply:
I need it to be called in this format `this.services_.register(key, service, opt[0], opt[1], opt[2], ...)` so I picked .apply() since it accepts all the arguments as a single array. There might be a way to do something like... `this.services.register(...arguments)` but I'm not sure -- and I seem to recall that gcc blocks developers from using the "arguments" pseudo-variable.